test/e2e/antreapolicy_test.go

```diff
@@ -2682,58 +2639,87 @@ func testAuditLoggingEnableNP(t *testing.T, data *TestData) {
+	k8sNPBuilder = k8sNPBuilder.SetName(namespaces["x"], npRef).
+		SetPodSelector(map[string]string{"app": serviceName}).
+		SetTypeIngress().
+		AddIngress(v1.ProtocolTCP, &p80, nil, nil, nil,
+			map[string]string{"pod": "a"}, nil, nil, nil)
 
-		var expectedNumEntries, actualNumEntries int
-		srcPods := []string{namespaces["x"] + "/b", namespaces["x"] + "/c"}
-		expectedLogPrefix := []string{npRef + " <nil> Allow [0-9]+ ", "K8sNetworkPolicy <nil> Drop <nil> "}
-		destIPs, _ := podIPs[namespaces["x"]+"/a"]
-		for i := 0; i < len(srcPods); i++ {
-			srcIPs, _ := podIPs[srcPods[i]]
-			for _, srcIP := range srcIPs {
-				for _, destIP := range destIPs {
-					// only look for an entry in the audit log file if srcIP and
-					// dstIP are of the same family
-					if strings.Contains(srcIP, ".") != strings.Contains(destIP, ".") {
-						continue
-					}
-					expectedNumEntries += 1
-					// The audit log should contain log entry `... Drop <ofPriority> <x/a IP> <z/* IP> ...`
-					re := regexp.MustCompile(expectedLogPrefix[i] + srcIP + ` [0-9]+ ` + destIP + ` ` + strconv.Itoa(int(p80)))
-					if re.MatchString(stdout) {
-						actualNumEntries += 1
-					} else {
-						t.Logf("Audit log does not contain expected entry from %s (%s) to x/a (%s)", srcPods[i], srcIP, destIP)
-					}
-					break
-				}
-			}
-		}
-		if actualNumEntries != expectedNumEntries {
-			t.Logf("Missing entries in audit log with K8s NP: expected %d but found %d", expectedNumEntries, actualNumEntries)
-			return false, nil
-		}
-		return true, nil
-	}); err != nil {
-		t.Errorf("Error when polling audit log files for required entries: %v", err)
+	knp, err := k8sUtils.CreateOrUpdateNetworkPolicy(k8sNPBuilder.Get())
+	failOnError(err, t)
+	failOnError(waitForResourceReady(t, timeout, knp), t)
+
+	// generate some traffic that wget the nginx service
+	var wg sync.WaitGroup
+	oneProbe := func(pod *v1.Pod) {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			data.runWgetCommandFromTestPodWithRetry(pod.Name, pod.Namespace, pod.Spec.Containers[0].Name, serviceName, 5)
+		}()
 	}
-	failOnError(k8sUtils.DeleteNetworkPolicy(namespaces["x"], "allow-x-b-to-x-a"), t)
+	oneProbe(podXA)
+	podXB, err := k8sUtils.GetPodByLabel(namespaces["x"], "b")
+	if err != nil {
+		t.Errorf("Failed to get Pod in Namespace x with label 'pod=b': %v", err)
+	}
+	oneProbe(podXB)
+	wg.Wait()
+
+	srcIPs := append(podIPs[namespaces["x"]+"/a"], podIPs[namespaces["x"]+"/b"]...)
+	var expectedLogPrefix []string
+	for range podIPs[namespaces["x"]+"/a"] {
+		expectedLogPrefix = append(expectedLogPrefix, npRef+" <nil> Allow [0-9]+ ")
+	}
+	for range podIPs[namespaces["x"]+"/b"] {
+		expectedLogPrefix = append(expectedLogPrefix, "K8sNetworkPolicy <nil> Drop <nil> ")
+	}
+	destIPs := serverIP.ipStrings
+	checkAuditLoggingResult(t, data, serverNode, "K8sNetworkPolicy", srcIPs, destIPs, expectedLogPrefix)
+
+	data.deleteService(service.Namespace, service.Name)
+	nginxCleanupFunc()
```

Comment:
they should be deferred right after the resources were created successfully, otherwise they won't be executed on other failures.

Reply:
deferring cleanup func, and added fail on error for deleting service.